Show image history in details panel

strzinek opened the pull request:
---
podman: Show image history

Image details now contain the full view of history an image.

![Screenshot from 2022-06-07 13-44-01](https://user-images.githubusercontent.com/12330670/172371389-bce80d41-6676-44d1-9306-3d9ae4b017b0.png)

strzinek commented:
>     1. When there's space (desktop sizes), use the horizontal space for the date and whatever text is there, currently next to the date. But... probably keep it stacked vertically on mobile.
>        
>        * They could, alternately, be headers and only shown when they change. As they're mainly the same across commands (dates, same text), this could work.
> 
>     2. Headings should be before the commands, not after. (To the left in desktop, above in mobile.)
> 
>     3. Commands probably need to look like commands, complete with with a monospace font. This shouldn't really be a standard pre element, but should have pre-wrap applied to it in CSS.
> 
> 
> I would have recommend a [PatternFly log viewer](https://www.patternfly.org/v4/extensions/log-viewer), however it only supports one field, and we need more than what it provides here. (Same for our system logs.)

I reworked the panel with ListingsTable component and ClipboardCopy expandable and added history numbering and data sizes.

garrett commented:
Thoughts:

- Do we need to have a copy and expand function? Can't we just use a simple monospaced font (that wraps)? I don't think copying the text is going to be a normal use. (Selecting and copying the text should still work of course.)

- Should we use a table at all? If we do, we do have to make things work in mobile mode. Right now, it doesn't. (See mobile screenshot below.)

- At least in the containers I've seen, the comments are useless. They're often completely blank or a "built by some_tool" message. I guess some people might actually use them for something, however.

- Do we want to show disk space? As you can't really interact with each line, it's not super-useful, but I could see it being interesting when running a command that happens to pull in a hundreds of megabytes of packages unexpectedly, when someone is specifically building an image on their own machine. (Otherwise, seeing the size when using built images from others isn't so interesting.)

Perhaps we can use a table, but simplify the entry and make sure headers don't show (especially on mobile) and do some special formatting.

I do have something in mind and am happy to make a mockup for it, but I ran out of time for today. (It's not so far off from what you have actually, in desktop mode, at least... Mobile is a different story though.)

I think all we need now are a few tweaks to get this PR in a state where everyone's happy.

---

And here are screenshots to show you what I'm basing this on (as what I see with some containers may be different form what you're seeing in some cases):

Here's a "typical" container from docker hub:

![image](https://user-images.githubusercontent.com/10246/165557600-b5e1cd59-3af7-48ad-92e4-e3b205014178.png)


What it looks like with no comments (debian image from docker hub):

![image](https://user-images.githubusercontent.com/10246/165557311-e193dcda-8304-4d06-9540-d9047f44c172.png)

Here's what toolboxes and intermediate images look like:

![image](https://user-images.githubusercontent.com/10246/165562488-8ba7cc35-d6f6-4866-ab3e-dbcb0b968bcb.png)


Mobile (it's pretty bad here, especially since the content itself is not even visible... and the labels are mostly unnecessary, yet take up the most space):

![Screen Shot 2022-04-27 at 17 28 15](https://user-images.githubusercontent.com/10246/165557177-5ebb5b74-3a27-4cd9-ad2e-65d83d6f0720.png)

strzinek commented:
> Perhaps we can use a table, but simplify the entry and make sure headers don't show (especially on mobile) and do some special formatting.

I simplified it a bit and removed the comment. I do not know hot to hide headers of cockpit ListingTable component when in narrow mode.

strzinek commented:
I managed to improve narrow history layout with only css, @garrett, have a look at it please.

garrett reviewed (changes requested):
Oh! I started a review a few days ago and just realized I didn't hit submit. Apologies!

strzinek commented on src/Images.scss:
```diff
@@ -0,0 +1,46 @@
+@import "global-variables";
+
+#containers-images div.download-in-progress {
+    color: grey;
+    font-weight: bold;
```
This is actually not my code, I just added stuff for image history and changed it from css to scss, but I'll try to fix it anyways ;-)

strzinek commented on src/Images.scss:
```diff
@@ -0,0 +1,46 @@
+
+#containers-images .pf-c-table.pf-m-compact .pf-c-table__action {
+    --pf-c-table__action--PaddingTop: 0.5rem;
+    --pf-c-table__action--PaddingBottom: 0.5rem;
```
All comments on style addressed, please review.

garrett commented:
I've made a simple commit that brings over Cockpit's change of dropping the gutter in mobile sizes, which should help ease the layout a little in mobile sizes, in PR https://github.com/cockpit-project/cockpit-podman/pull/990. It should help out here a bit.

I'll re-review this PR now. Thanks for the changes!

garrett reviewed (changes requested):
I've left a lot of comments on the CSS, and meant to address them locally... but I got sidetracked. Apologies!

I'm about to start a long weekend and next week is the Cockpit team meeting week, so I'm not sure when I can come back to this, but I'd like to ASAP. Perhaps I can squeeze in some time next week to make some of these suggested changes.

I _did_ make a lot of these changes that I will push up though. It's not done, but a lot of these changes have been made.

strzinek commented on src/client.js:
```diff
@@ -162,9 +162,18 @@ export function getImages(system, id) {
                             })
                             .catch(reject);
+
+                    for (const image of immages || []) {
```
No, it is not a typo. Somebody before me defined such name to distinguish between results from API call and a list with supplemented details. I just used it too.

strzinek commented on src/Images.scss:
```diff
@@ -0,0 +1,46 @@
+
+@media ( max-width: $pf-global--breakpoint--md ) {
+
+    .pf-m-grid-md.pf-c-table [data-label] > .ct-code {
```
I need to target a fairly embedded column to override PF generated style, so yes, it needs to be specific.

marusak commented:
@strzinek Can you please rebase to fix conflicts?

marusak commented:
@strzinek Why did you close this PR?

strzinek commented:
I screwed up the rebase of pl branch, and close was just the result of force push. I´ll get back to it later.

strzinek commented:
So, I fixed up the branch, hope it will be ok now.

marusak commented on src/ImageDetails.jsx:
```diff
@@ -34,7 +34,7 @@ const ImageDetails = ({ containers, image, showAll }) => {
             </DescriptionListGroup>
             }
-            {image.Ports.length !== 0 &&
+            {image.Ports && image.Ports.length !== 0 &&
```
Is this related? have you seen this failing that `Ports` were not defined? If so, it should be a separate commit (possibly separate PR).

marusak reviewed (changes requested):
Looks great, thank you!

Some tiny issues but mainly we need test for this. We `podman build` in tests so we could use that for example.

marusak commented on src/client.js:
```diff
@@ -166,9 +166,18 @@ export function getImages(system, id) {
                             })
                             .catch(reject);
+
+                    for (const image of immages || []) {
```
This cannot be like this here as it may execute and `resolve()` before previous promises are done, thus we might miss some information. You need to do this in sync with previous promises.

jelly replied:
This also seems to be a performance issue when loading images, so we should either use:

https://docs.podman.io/en/latest/_static/api.html#tag/images/operation/ImageListLibpod which already has History but lacks the image size. Or we load the History on demand (which makes a lot more sense) when opening the tab. We do the same thing with the Container integration tab.

marusak replied:
Yes, I think on demand would be the best option. Storing history for each image in memory is rather useless as I don't expect a load of people looking into it. And waiting for second when toggling the tab seems like ideal solution.

jelly commented:
Implemented in https://github.com/cockpit-project/cockpit-podman/pull/1074
